### ISSUE TITLE
Port to WebExtension

### ISSUE THREAD

**Issue by yodaldevoid** [denschub/firefox-tabgroups#60]
With the "old" plugin formats for Firefox getting support dropped not terribly long into the future, will this plugin survive? The other big tab groups plugin has had it's developer decide to stop supporting it as the amount of work to port it to WebExtensions is too great for what he gets out of it. Will development support for this also be stopped, or will it be ported to WebExtensions? No matter how this pans out, thank you for all the work on this plugin.

**Comment by skyschub** [comment]
> will this plugin survive

yes.

**Comment by yodaldevoid** [comment]
Sorry for the melodrama, it may have been a little too late (early?) when I submitted this.

Beyond that, I would love to help port this or do any other work to make that easier. I'm not a plugin developer by any means, so if there are sub-tasks to porting the plugin please feel free to list them as I would not know what they are otherwise.

**Comment by teranex** [comment]
I'm currently using [Tab Groups](https://addons.mozilla.org/en-US/firefox/addon/tab-groups-panorama/?src=search), but this will not receive updates anymore, so I'd like to give this extension a try. If I install this extension, will it use all the tabs and tabgroups I currently have in Tab Groups? And will I be able to switch back from this extension to tab groups, should I feel the need?

**Comment by yodaldevoid** [comment]
The tab groups transfer between the two addons without needing to do anything. For questions like this in the future please create a new issue.

**Comment by ScottRFrost** [comment]
I noticed that there is no Project or Milestone set for this issue.  Any idea if porting to web extension will happen before [Firefox 57 hits beta on 2017-09-26, when this addon will cease to function](https://blog.mozilla.org/addons/2016/11/23/add-ons-in-2017/)??

**Comment by machineloop** [comment]
Thanks @denschub for your work on this add-on over the years, I've really enjoyed using it. Nightly (57.0a1 2017-08-11) just disabled legacy add-ons 6 weeks ahead of release of 57. I re-enabled legacy add-ons, but looks like the button doesn't show up anymore to access tabgroups. Anything I can do to help? I've done a bit of work with Firefox add-ons in the past, but nothing with the new WebExtension APIs, but would be happy to pitch in.

**Comment by skyschub** [comment]
Read [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1384515) and the comments on that. I will start working on this as soon as the API landed, just as I promised. Right now, there is nothing I can do and there is nothing I will do, since I'd rather have something clean than a a hack that I need to update every 2 days to have it working.

A lot of stuff will change in Nightly over the next few weeks, and I will be busy with work, so I'm not going to maintain a hack. I will, however, port the extension as soon as I have the APIs I need.

**Comment by scheinercc** [comment]
@machineloop have you found a way to access those tabs?

**Comment by skyschub** [comment]
To restore hidden tabs:

1. Open the DevTools (F12), open the DevTools settings (the gear icon in the top right corner of the Devtools), enable `Enable browser chrome and add-on debugging toolboxes`.
2. Open the Browser console (Tools -> Web Developer -> **Browser console**).
3. Paste `gBrowser.showOnlyTheseTabs(Array.from(gBrowser.tabs))` into the input area and hit enter.

**Comment by gerhardj** [comment]
It seems, there is some progress with that new API for hiding tabs, right?

**Comment by edouard-lopez** [comment]
yep there is some[ experiment going on](https://bugzilla.mozilla.org/show_bug.cgi?id=1384515#c65)

**Comment by grahamperrin** [comment]
For the benefit of anyone who is not closely following things in [above-mentioned](https://github.com/denschub/firefox-tabgroups/issues/60#issuecomment-321878075) Mozilla bug [1384515 - Provide an API for hiding and showing individual tabs](https://bugzilla.mozilla.org/show_bug.cgi?id=1384515), two novelties of particular interest:

- [WebExtensions/TabHiding - MozillaWiki](https://wiki.mozilla.org/WebExtensions/TabHiding)
- Mozilla bug [1408053 - Show that tabs are hidden and all the implications](https://bugzilla.mozilla.org/show_bug.cgi?id=1408053)

– please discuss in the appropriate areas.

**Comment by skyschub** [comment]
> please discuss in the appropriate areas.

Please don't duplicate information. If people are interested in the Firefox development side, they can follow 1384515. All new issues are linked in there. If people decide not to follow the issue on bugzilla, than that's because they are not interested in the development history and only want to know when my extension is ported.

**Comment by grahamperrin** [comment]
Apologies. Deleted.

**Comment by rubcap** [comment]
Thanks @denschub for your continued work on this add-on!  I understand this won't make it to Firefox 57 due to a missing WE-API, but I really look forward to a WE compatible version! 👍

**Comment by nkkollaw** [comment]
I'm having somewhat of a hard time figuring out the status of this. Is this extension getting ported to WE?

I suck and wouldn't know where to start with helping, but I do enjoy, love, and need this extension and would hate to see it die.

**Comment by skyschub** [comment]
> Is this extension getting ported to WE?

Oh please... ;)

<img width="414" alt="screen shot 2017-11-10 at 02 27 44" src="https://user-images.githubusercontent.com/344777/32638257-c1cdbe94-c5be-11e7-9103-1603853deb4e.png">

As mentioned before: I will update this issue as soon there is something to update.

**Comment by skyschub** [comment]
So, before I get another 91 emails (yes, I'm serious) to inform me that tabHide support has landed: *I know*. Hope to find the time to spend some time coding here soon, and I will update this thread as usual.

**Comment by skyschub** [comment]
Alright people. It's more than a year since this issue has been created, and I think I owe you all a reply.

Back in April last year, I assumed that landing required WebExtension APIs should be done very soon, and made time for that in my schedule. As everyone who followed the Bugzilla thread is aware, implementing a tabHide API actually was much more complicated than anticipated, and thus got delayed by quite a bit.

While I had the time and motivation to port this extension to WebExtension last year, I've since moved on. I depend on using Firefox Nightly on a daily basis, so I was unable to use tab groups for quite some time now, and today, I don't really have the need for it anymore. And as I don't need it for myself anymore, I cannot really justify the time investment to port it these days. There is just too much stuff going on in my life, y'all know how it is.

So, long story short: I'll put this repository in an "officially unmaintained" state, as I won't be able to put the effort in anytime soon.

That being said, if anyone wants to step forward and do the port, feel free to do so. I am very happy to mentor an implementation and answer every questions, feel free to [reach out to me](https://schub.io/contact) in that case. If there is a working port, I'm more than happy to transfer this repo and also the AMO listing to the new maintainer.

Harsh decision I have to make, but I'm sure most of you understand. Thanks for the support!

**Comment by hadrianw** [comment]
I would only like to point people to alternatives. I currently use [Totally not Panorama](https://addons.mozilla.org/en-US/firefox/addon/basic-panorama/) which itself is a fork of [Panorama View](https://addons.mozilla.org/en-US/firefox/addon/panorama-view/). They are almost a complete copy of the original Tab Groups feature. So they're not Simplified, but more Full Tab Groups.

Totally not Panorama works great for me - it supports tab hiding.

**Comment by Wobbler44** [comment]
Not familiar with 'Totally not Panorama', but I recently started using 'Sync Tab Groups' by Morikko which works nicely and was finally able to do the long outstanding Firefox update and abandon this legacy 'Simplified Tab Groups'  - thanks Dennis for your past work.

**Comment by grahamperrin** [comment]
@denschub thanks for the frank explanation 👍

> … implementing a tabHide API …

Shortly before your comment, I began searching Bugzilla for anything relating to a **groups** API. Can't find one … am I missing something? https://discourse.mozilla.org/t/-/28533?u=grahamperrin if you'd like to reply there. TIA

----

All: https://github.com/Quicksaver/Tab-Groups/issues/535 seems to be a favoured meeting place for developers, testers and users of possible alternatives to _Tab Groups_. Maybe use the same issue 535 for discussion of possible alternatives to _Simplified Tab Groups_ (prefer to track the mixture in one place, not two or more). Cheers.

**Comment by skyschub** [comment]
@grahamperrin you don't need a "groups" API, and even pre-quantum, there was none. You have to keep track of the tabs yourself and hide/show them when needed.

**Comment by bz007** [comment]
@hadrianw thanks a lot for TnP! It is exactly what I wanted!

**Comment by psxlover** [comment]
Can anyone who has tried the latest versions of Panorama View and Totally not Panorama post a comparison? Thanks

**Comment by hadrianw** [comment]
@psxlover: from Totally not Panorama extension's description:

> Enhancements, compared to Panorama View 0.6.15 by photodiode:
> - Support for hiding tabs (Firefox 59+)
> - On-demand screenshots (Firefox 59+)
> - Support pinned tabs
> - Ctrl+E opens/closes the panorama view (Ctrl+Shift+E for Mac)
> - The keyboard shortcut is configurable for Firefox 60 users. You can only set shortcuts which are not otherwise used by the browser.
> - Dragging a tab out of a group creates a new group with this tab in it
> - Resize and move groups
> - Create a new group by drawing in the empty space in the panorama view
> 
> Limitations (compared to Tab Groups by Quicksaver):
> - Doesn't work in Private Browsing windows
> - No "Undo closed group" button
> - No automatic group management - you need to assign tabs to groups manually
> - No context menu actions
> - Groups may overlap, you need to select appropriate sizes manually

**Comment by psxlover** [comment]
I saw that, but that is referring to version 0.6.15 of Panorama View which is old. From what I understand Panorama View does support hiding of tabs, and has had several updates since the last version of Totally not Panorama was released.

**Comment by Termuellinator** [comment]
@psxlover , i just tried TnP now, have used PV for a while - for now, i like TnP better because the handling of pinned tabs suits me better and it feels a tad smoother when changing groups.
All in all they are both very similar and great addons and a good replacement for tab groups :)

**Comment by psxlover** [comment]
> the handling of pinned tabs suits me better
How are they handled in each addon? Which one is the same as the old tab groups (pinned tabs stay visible on all groups), because I use pinned tabs for e-mail and I want them always visible.

**Comment by emyarod** [comment]
@psxlover @hadrianw for what it's worth: [the developer of Totally not Panorama now contributes to Panorama View](https://www.reddit.com/r/firefox/comments/84ouqc/panorama_view_an_aboutconfig_tweak_in_is_the/dvuu1xq/)

**Comment by Termuellinator** [comment]
@psxlover in PV they are viewn as part of a group, but when you click them while in another group they will appear on their own - kinda weird in my opinion.
in TnP they are viewn seperately on the right and appear in the tab-bar of every group.
@emyarod interesting - we'll have to wait and see how things go further O_o

**Comment by psxlover** [comment]
@Termuellinator I just tried Panorama View, and even though the pinned tabs appear as part of group, they work correctly when in another group (they show while having selected another group, and they don't switch the tab group when selected)

**Comment by ScottRFrost** [comment]
[Conex](https://addons.mozilla.org/en-US/firefox/addon/conex/) is the closest I've found to a replacement.  It's not as slick as TabGroups was, but it does have the additional functionality of integrating with Firefox Multi-Account Containers (so sessions / cookies / etc stay in the "group" and can't be seen in other groups). 

It works pretty well, but I'd love to see a true TabGroups for WebExtensions.

**Comment by Keith94** [comment]
>It works pretty well, but I'd love to see a true TabGroups for WebExtensions.

@ScottRFrost The closest you will get is [Simple Tab Groups](https://addons.mozilla.org/addon/experimental-simple-tab-groups/), the one I use personally. :-)

**Comment by alex29ua** [comment]
Anything that has the tabs panel for groups as well? So I don't have to click the button to see the groups like in "Sync Tab groups" etc. They all don't have screenshots, hard to choose...

**Comment by alex29ua** [comment]
Anything that has the tabs panel for groups as well? So I don't have to click the button to see the groups like in "Sync Tab groups" etc. They all don't have screenshots, hard to choose...

**Comment by pinta83** [comment]
@alex29ua Simple Tab groups from two reply's above.... You have simple groups, but you can click on the "Manage groups" to get almost identical grid view to old Tab Groups... Alltough if you name your groups right and setup colors to your needs - you won't have to go into grid/manage view

**Comment by alex29ua** [comment]
I don't need anything more complicated than a simple tabs row. Like I already have for usual pages, but for groups. 
This way I always see the groups and counts. They are like contexts/topics to me, easily switchable. Also easy to move tabs/pages around. 

Possible in one of replacements?
![image](https://user-images.githubusercontent.com/6235825/40052878-dc667f68-5847-11e8-92d3-7b199ad01b05.png)

**Comment by alreadydone** [comment]
In the old TabGroups Manager you can drag one or multiple tabs to another tab group, a feature that seems to missing in nowadays tab group WEs. (Tab groups are listed horizontally and tabs can be listed vertically through add-ons.)

**Comment by gaekwad** [comment]
I tried TnP and it's pretty good.  The only thing I don't like about it is
there doesn't seem to be shortcut keys for switching to the next/previous
group like there was in the old Tab Groups.  I switched groups a lot, and
now to do that I'd have to do Ctrl+E (or whatever I have it set to) and
click on a tab every time.

**Comment by rustymonkey** [comment]
I've looked at all of the alternatives suggested above and all require the permission "Access your data for all websites". Can anyone who's developed these types of addons explain whether this is actually necessary for an implementation of a Tab Groups type of feature? I'm a bit apprehensive of giving a third-party addon this level of access. Thanks.

EDIT: it looks like Morriko's 'Sync Tab Groups' doesn't require this permission. Not sure why all the others do...

**Comment by peci1** [comment]
@rustymonkey As I understand it, the creators of these addons usually require it for fetching favicons or page previews. If it's the only reason, it should be possible to create an addon that doesn't require this privilege, but the UI would be less user-friendly.

**Comment by fturcheti** [comment]
I have tested all the alternatives above and couldn't really stick to none. At the moment I'm using [Tiled Tab Groups][1], by chzesa, and it fit my needs better.

[1]: https://addons.mozilla.org/en-US/firefox/addon/tiled-tab-groups/

**Comment by shangles** [comment]
Totally not Panorama works pretty good for me.

**Comment by sacp** [comment]
> Totally not Panorama works pretty good for me.

That's also what I've been using for some weeks now

**Comment by Vampire** [comment]
> I currently use Totally not Panorama which itself is a fork of Panorama View. They are almost a complete copy of the original Tab Groups feature. So they're not Simplified, but more Full Tab Groups.

Can someone say how those compare to this add-on?
Was this more elaborate, or was this more simplified, or was it just different.
It's so long that no tab groups worked, that I forgot why I used this one and not one of the others and what the features were exactly.

**Comment by thePanz** [comment]
I found TabPanorama hanging when loading lot of tabs (imported from TabGroups)
Panorama does *not* support lazy loading of tabs
It is quite hard to manage groups with panorama, the list-view of this Simple-TabGroups is way better IMHO

**Comment by archerindigo** [comment]
Tried Simple-TabGroups and Panorama View for some time.

Simple-TabGroups use the old list-view which is more comfortable but it reload all the tabs when switching group. That's not what I desired.

Panorama View is better to view all the tabs opened but it doesn't support tab-hiding and not so smart in ordering and grouping the tabs. It seems to be RAM hungry either as I noted the RAM usage raise quickly during the days using it.

Both are not the perfect replacement IMHO.

**Comment by aullik** [comment]
[Panorama View](https://addons.mozilla.org/en-US/firefox/addon/panorama-view/) has switched over to tab hiding a while ago. It allows orienting the tab-groups freely. 

However it is RAM hungry ~~and it has no export/import functionality yet.~~

Panorama View is more of a replacement to the old Quicksaver/Tab-Groups that died with FF57

Thanks @udovdh for correcting me. The import export function can be found under 
about:addons => Panorama View => Options

**Comment by udovdh** [comment]
Panorama view _has_ export/import but with issues.
RAM usage I did not quantify.

**Comment by Vampire** [comment]
Didn't this plugin also have the `Ctrl+E` view and not only a list view?
I think I used this until the end (WebExtensions only).

**Comment by udovdh** [comment]
There is a panorama-type overview in Panorama view

**Comment by Vampire** [comment]
As is in TnP as far as I understood and as was in this add-on as far as I remember.

**Comment by archerindigo** [comment]
> Panorama View has switched over to tab hiding a while ago. It allows orienting the tab-groups freely.
>
>However it is RAM hungry and it has no export/import functionality yet.
>
>Panorama View is more of a replacement to the old Quicksaver/Tab-Groups that died with FF57
>
>Thanks @udovdh for correcting me. The import export function can be found under
>about:addons => Panorama View => Options

Alright I just forgot to enable tab hiding

**Comment by thePanz** [comment]
@archer1609 SimpleTagGroup implemented the lazy-loading of tabs. Tabs are displayed but not "running" until you activate them.

**Comment by archerindigo** [comment]
@thePanz 

I personally don't really like it. Usually I would like to keep the tabs loaded even if I switch to another group. And this kind of lazy-loading takes quite a long time to fully switch to another group.
Thus, Panorama View might suits me better.

**Comment by itzallan** [comment]
I've been using Tree Tabs for several months and have been happy with it so far. I like how it allows me to create groups and drag tabs into groups.  It also lets me create folders within groups so that I can further organize my tabs.  I also use Tab Suspender to automatically unload tabs after a certain amount of time of inactivity to help economize on memory.

**Comment by mibocs** [comment]
Is there a plugin that can import tabs from tabgroups automatically?

**Comment by me21** [comment]
Simple Tab Groups. It has implemented tab hiding now, so it doesn't have to reload them on switching the group. But this is configurable.
Just make a backup in this extension and import it in STG.

**Comment by mibocs** [comment]
Cool, thanks!

**Comment by mibocs** [comment]
@me21 What do i have to do to hide/unload my tabs? So i dont't want my tabs to be loaded everytime, just only when i click on it. There is a option called "Discard tabs after hide (this will save the RAM)". Do i have to just enable this option? Or do i have to install another extension like this: https://addons.mozilla.org/en-US/firefox/addon/auto-tab-discard ?

**Comment by ThomasHeinrichSchmidt** [comment]
I'm also using "Simple Tab Groups" now, works smoothly just as you would expect it. I've not checked the option "Discard tabs after hide (this will save the RAM)", and my tabs are only loaded when I click on them (after changing to a different Tab Group).

**Comment by me21** [comment]
@mibocs Tabs should always be loaded only when you click on them. Or just shown if they're not discarded.

**Comment by mibocs** [comment]
@me21 yes, this is not necessary now, but what does this option do? How can you hide tabs? There should be also such a option for auto unload when you don't use tabs for a while these should be unloaded to save memory. This plugin https://addons.mozilla.org/en-US/firefox/addon/auto-tab-discard does exactly what i say.

**Comment by me21** [comment]
@mibocs I think when you switch to the new tab group, tabs from the old group are hidden. Or unloaded, if the option to discard tabs is enabled.

I think you can use the plug-in you're referring to with Simple Tab Groups just fine. STG itself hides/unloads tabs only when you switch groups.

I am not the developer of Simple Tab Groups, though, and I would advise you to ask questions and post suggestions to Simple Tab Groups GitHub repository.

**Comment by mibocs** [comment]
@me21 https://github.com/Drive4ik/simple-tab-groups/issues/201 it seems when discard option is enabled it unloads all tabs from inactive groups.

**Comment by mibocs** [comment]
I asked "Discard tabs after hide (this will save the RAM)" option here https://github.com/Drive4ik/simple-tab-groups/issues/251, i hope we will get answer soon.

**Comment by me21** [comment]
@mibocs Yup, I think you're right. It unloads the tabs from inactive groups.

**Comment by mibocs** [comment]
The answer from @Drive4ik: "Hi! Yes, you and @me21 right. If this option is enabled and you change group from one to other - tabs for closed group will make as "discarded" and unload its from computer RAM"

But "Auto Tab Discard" is better because it discards only the inactive tabs and if someone changes between groups frequently they don't want all tabs to be discarded/unloaded but only the unused tabs. So in this case "Discard tabs after hide (this will save the RAM)" option can be disabled.

**Comment by Drive4ik** [comment]
@mibocs currently this option is disabled by default :)

**Comment by HeRo002** [comment]
Today Ubuntu updated Firefox ESR to v.60.2.1esr, disabling Tab Groups
etc., and leaving me with a total mess.
It seems I am only seeing the tabs from the group that was active before
the update (plus a few more tabs...?!).
Does anyone have any idea what to do now?
There is a message https://support.mozilla.org/en-US/questions/1232522
from someone with the same problem, but I can't make much sense of that
thread...(?)
I also had Session Manager and TabMixPlus installed, so I presume I have
several good session backups. Maybe they could be imported into Simple
Tab Groups (if I install that)?
I hope someone can help me, please!
I REALLY think Mozilla is screwing us big time!!

- Henrik

**Comment by SuperTux88** [comment]
See https://github.com/denschub/firefox-tabgroups/issues/60#issuecomment-322005792 for how to restore/show hidden tabs.

**Comment by HeRo002** [comment]
A few hours ago I wrote the following. Why on Earth did someone mark it as resolved??:

Today Ubuntu updated Firefox ESR to v.60.2.1esr, disabling Tab Groups etc., and leaving me with a total mess. 
It seems I am only seeing the tabs from the group that was active before the update (plus a few more tabs...?!). 
Does anyone have any idea what to do now? 
There is a message https://support.mozilla.org/en-US/questions/1232522 from someone with the same problem, but I can't make much sense of that thread...(?) 
I also had Session Manager and TabMixPlus installed, so I presume I have several good session backups. Maybe they could be imported into Simple Tab Groups (if I install that)? 
I hope someone can help me, please! 
I REALLY think Mozilla is screwing us big time!!
 - Henrik

**Comment by skyschub** [comment]
@SuperTux88 already answered your question, and thus, it is resolved. Read the linked reply and it will tell you how to get your tabs back. What you think about the move to WebExtensions API is your thing, but this issue is not the place for you to post your opinion. I am being more than liberal by allowing everyone here to discuss alternative extensions and alternative tab management approaches, but I will not let you or everyone else use this as a platform to shout out your opinion in that manner.

**Comment by HeRo002** [comment]
OK... So  SuperTux88  referred me to this post: 

" To restore hidden tabs:
    Open the DevTools (F12), open the DevTools settings (the gear icon in the top right corner of the Devtools), enable Enable browser chrome and add-on debugging toolboxes.
    Open the Browser console (Tools -> Web Developer -> Browser console).
    Paste gBrowser.showOnlyTheseTabs(Array.from(gBrowser.tabs)) into the input area and hit enter. "

Correction: 
I tried that. It DID make the list of tabs much longer - yes! But I did not get all my tab groups back. So the issue is NOT resolved!
